internal/container/non_root_container_check.go

```diff
@@ -0,0 +1,104 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License, Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ******************************************************************************/
+
+package container
+
+import (
+	"fmt"
+	"regexp"
+
+	txqualitychecks "github.com/eclipse-tractusx/tractusx-quality-checks/internal"
+)
+
+type NonRootContainer struct{}
+
+// NewNonRootContainer returns a new check based on txqualitychecks.QualityGuideline interface.
+func NewNonRootContainer() txqualitychecks.QualityGuideline {
+	return &NonRootContainer{}
+}
+
+func (n NonRootContainer) Name() string {
+	return "TRG 4.03 - Non-root container"
+}
+
+func (n NonRootContainer) Description() string {
+	return "Container images shall not run as root for security reasons."
+}
+
+func (n NonRootContainer) ExternalDescription() string {
+	return "https://eclipse-tractusx.github.io/docs/release/trg-4/trg-4-03"
+}
+
+func (n NonRootContainer) Test() *txqualitychecks.QualityResult {
+	checkPassed := true
+	var errorDescription string
+	dockerfiles := findDockerfilesAt("./")
+
+	for _, dockerfilePath := range dockerfiles {
+		file, err := dockerfileFromPath(dockerfilePath)
+		if err != nil {
+			fmt.Printf("Could not read Dockerfile from path %s\n", dockerfilePath)
+		}
+
+		if !validateUser(file.user()) {
+			checkPassed = false
+			errorDescription = "Invalid user specified in Dockerfile"
+		}
+	}
+
+	return &txqualitychecks.QualityResult{Passed: checkPassed, ErrorDescription: errorDescription}
+}
+
+func (n NonRootContainer) IsOptional() bool {
+	return false
+}
+
+// validateUser validates USER instruction in Dockerfiles and return a bool if a valid USER has been found.
+// To return true username/group-name must not be root or contain upper case letters, or must not be 0 or greater 65536
+// for uid/gid.
+func validateUser(u *user) bool {
+	if u == nil {
+		return false
+	}
+
+	validUser, _ := regexp.Match(
+		`(^(\d|[1-9]\d{1,3}|[1-5]\d{4}|6[0-4]\d{3}|65[0-4]\d{2}|655[0-2]\d|6553[0-6])$)|(^[a-z_][a-z0-9_-]*[$]?$)`,
```

Comment:
The regex pattern could be extracted into a constant to be reusable.

Reply:
Good idea. Done